torch/profiler/profiler.py

```diff
@@ -0,0 +1,191 @@
+import torch.autograd.profiler as prof
+from torch.autograd import ProfilerActivity
+
+from typing import Callable, Iterable, Optional
+
+class EnablePred(object):
+    """
+    EnablePred describes on which steps profiler is active:
+    - profiler starts in inactive state and stays in inactive state for the first 'wait' steps
+    - profiler then enters a warmup state and stays in this state for the next 'warmup' steps
+    - profiler then starts actively tracing/collecting stats for the next 'active' steps
+    - after this, profiler returns to the inactive state and cycle repeats
+
+    In case output_fn is specified, it is called every time the trace is ready
+    """
+    class Action(object):
+        START_WARMUP = 0
+        START_TRACE = 1
+        STOP_TRACE = 2
+
+    class State(object):
+        INACTIVE = 0
+        WARMUP = 1
+        ACTIVE = 2
+
+    def __init__(self, wait: int, warmup: int, active: int, output_fn: Optional[Callable[[prof.profile], None]]):
+        assert wait >= 0 and warmup >= 0 and active > 0
+        if warmup == 0:
+            print("Warning: profiler won't be using a warmup, which can skew profiler results")
+        self.wait = wait
+        self.warmup = warmup
+        self.active = active
+        self.output_fn = output_fn
+
+        def active_active_fn(step):
+            if self._mod_step(step) == 1:
+                return [EnablePred.Action.STOP_TRACE, EnablePred.Action.START_WARMUP, EnablePred.Action.START_TRACE]
+            else:
+                return []
+
+        def inactive_warmup_fn(_):
+            raise RuntimeError("Incorrect profiler state sequence")
+
+        self.actions_map = {
```

Comment:
I still feel it can be simplified. Can we replace the entire EnablePred with a function that returns one of the few enums of what to do on this iteration:
* NONE
* WARMUP (i.e. enable but not record)
* RECORD (i.e. enabled)
* RECORD_AND_FLUSH (i.e. enabled but dump a new iteration)

Then the profiler doesn't even need to depend on EnablePred class. It can be just an arbitrary callable with that maps `step_num` into `action`. Of course you'd provide the code you have here as the default implementation but the user can do their own. E.g. the simple one would be: `enable_pred = lambda num: WARMUP if num < 100 else RECORD` for simple profile that doesn't dump intermediate results.

This logic about state machine transition is best moved to Profiler directly

As for `output_fn` - what benefit does it have over just putting `if p.step % 100 == 0: <my code>` directly in the training loop? You can keep it, but I'd add it as a separate arg in addition to `enable_pred` so that both can be easily-passable lambdas

Reply:
Good discussion here and above - I agree we should strive to simplify this as much as possible, and then make it as easy to understand as possible via good names and examples.

Warning: I haven't read the entire code yet so I might be missing the point - my comments are about user experience, not so much implementation details.

In my experience, a default behavior is be good for the majority of cases so I'll use that as example: one warmup iteration followed by three recorded iterations.
What frequently changes is when you want to trigger it (should allow simple modulo approach here I think) and what to do with the result. Less frequently, how many iterations to profile and which activities to include.

Given that, the standard use case ought to be something like:
trigger_when: p.step % 1000 == 100
handle_result: write_chrome_trace / print_summary / extract_metrics .... (list of predefined convenience functions)

i.e. trigger profiling at iteration 100, 1100 and so on. Important to note here is that profiling cannot be enabled while at the same time reporting. We need to warmup, record, then stop and process, in order for the overhead to be manageable during the record window.

Reply:
To clarify the above, this will be the behavior:
1100: Warmup
1101-1102: Record
1103: Record, then stop and process, calling the handle_result function asynchronously when processing is complete. As much work as possible is deferred to the processing stage to minimize overhead during recording. Recording should not continue in parallel with processing for detailed profiling, but we also need to consider a light-weight continuous profiling use case, possibly streaming a trace or ongoing extraction of metrics.

Reply:
@dzhulgakov  suggestion sounds good, basically let's provide a way to define a sequence of states explicitly through lambda + suggested default impl

one caveat - we might then allow users to specify sequences like: warmup none warmup none, etc - so we'll need to check for that, but i guess that's fine - assuming most users use supplied default and the power users know what they are doing;

@gdankel I think both solutions above should basically allow these kind of use cases

Reply:
Sure... As long as users don't have to deal with this level of detail in the majority of cases, it seems reasonable.